### ISSUE TITLE
Reduction accumulators are localized only when the numerics policy asks for a widening, so every f32 reduction no schedule op reaches accumulates in global memory — and Metal's volatile RMW workaround pins it there

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#693]
Split out of [#683](https://github.com/ahrefs/ocannl/issues/683), whose `w_o` site is one instance. #683's own framing is a *matcher miss* — no matmul sketch is ever seeded there — and its fix is to seed one. This issue is the mechanism that made the miss expensive rather than merely unoptimal, and it survives #683's fix at every other site no schedule op reaches.

## The claim

**A reduction accumulator is moved out of its node's storage only when the numerics policy asks for a widening, or when a schedule op mints a scope.** There is no precision-neutral, schedule-independent localization. So an f32 reduction that no schedule op reaches accumulates *in the output node's global memory*, one read-modify-write per step, and the only thing standing between that and the hardware is the backend C/HIP/MSL compiler's own LICM store promotion.

Three facts pin it:

1. **The codegen peel that would localize it declines at identity precision.** `try_widen_serial_reduce` ([`arrayjit/lib/c_syntax.ml:3566`](https://github.com/ahrefs/ocannl/blob/master/arrayjit/lib/c_syntax.ml#L3566)) already *recognizes* the nest — `Low_level.peel_accum_nest` peels Serial/`Unrolled` levels, sees through pure-index guards, hoists through a scope-form base — and then bails on

   ```ocaml
   if Ops.equal_prec (acc_prec store_prec) store_prec then None
   ```

   Its own comment states the consequence: "Inert (`None`) where `acc_prec` is the identity on the target's storage precision — f32/f64/integers, `narrow_compute_f32 = false`, native fp16, and the GPU backends' 16-bit precisions whose tensor units accumulate at storage width — those render the plain serial loop unchanged." That is correct for gh-639's purpose (accumulator *width* is policy) and it is the whole of the mechanism: the rewrite is a widening pass that localizes as a side effect, not a localization pass.

2. **The IR-level route is closed by construction.** `Local_scope` in `LL.optimize` is virtualization-only — the inlined computation of a *virtual* node. A `Local_scope` over a materialized node is out of contract there and is silently normalized back to a plain `Get` ([#681](https://github.com/ahrefs/ocannl/issues/681)). So the accumulator of a materialized output cannot be localized before codegen even in principle, and codegen's one rewrite is the precision-gated one above.

3. **Everything else that localizes is schedule-driven.** The accumulation mints are `Sched.Unroll ~materialize:true`, `Partition`, `Split_reduce`, and `try_register_tile`'s C-tile / `try_vectorize_reduce`'s epilogue. All of them require the site to have been proposed, timed, and crowned. gh-639's principle — "a reduction's effective accumulation width is set by the numerics policy, never by which schedule happened to place the accumulator in a register" — was stated for *width*. **Residency was never given the same treatment**: at f32 it is still exactly "whichever schedule happened to place it".

## Why it is not merely a source-level cosmetic

On cc / CUDA / HIP the per-node pointers carry `restrict` / `__restrict__` (`C_syntax`'s `restrict_keyword`, gh-164), so LLVM's LICM store promotion *can* hoist a loop-invariant-address accumulator into a register. Whether it does at any given site is unverified — I have not read the ISA for #683's `w_o` kernel, and #683's "that alone plausibly accounts for most of the 9.2%" should be read as the hypothesis it is; 1024 resident threads is the competing explanation and may well be the dominant one.

**On Metal it provably does not.** `volatile_scalar_rmw = true` ([`arrayjit/lib/metal_backend.ml:568`](https://github.com/ahrefs/ocannl/blob/master/arrayjit/lib/metal_backend.ml#L568)) shadows the node's pointer with a `volatile`-qualified alias for exactly this statement shape, precisely to stop the shader compiler from touching it — a miscompilation workaround (`test/operations/scalar_rmw_accumulation.ml`, repro in `benchmarks/runners/ocannl/bench_metal_bug.ml`). Its trigger predicate is verbatim the localization opportunity:

> a read-modify-write whose address is invariant across at least one enclosing serial `for` loop (a scalar loss reduction's constant index, a gradient accumulated over an outer batch loop, a matmul/conv accumulator indexed only by loops outside its reduction)

So on Metal, every un-scheduled f32 reduction — every cross-entropy loss reduction, every gradient accumulated over a batch loop, every unmatched contraction — performs a genuine volatile load+store to device memory per accumulation step, guaranteed, by design.

There is a second prize in that: **localizing the accumulator removes the RMW that the workaround exists to protect.** The scope form stores once, after the nest. Whatever fraction of sites a precision-neutral peel covers is a fraction where the volatile shadow no longer has anything to fire on.

## What a fix touches

- Relax the identity-precision gate in `try_widen_serial_reduce` from "widening needed" to "recognized accumulation nest", keeping the widening behavior as the special case. The rewrite it performs is already value-neutral at storage precision — the `Scope` arm's hoist comment says so explicitly ("at storage precision the hoist is value-neutral") — so this is a gate change, not new emission.
- Decide the interaction with `volatile_scalar_rmw`: once the RMW is gone at a site the shadow should not fire, and the `reads_tn` / `body_reads_tn` walk currently still sees the scope body's init read.
- The existing declines stay: `debug_log_from_routines` (a scope body renders with `log_set_locals:false`), and `mentions_rng_conversion` (gh-517).
- Watch for the cases where the gate is currently load-bearing beyond precision — a materialized node whose partial value must be visible to a sibling statement mid-nest is not a peelable accumulation, but that is `peel_accum_nest`'s job and it already refuses a sibling statement at a level.

## How to measure it

The cheap discriminating experiment is Metal, where the outcome is not at the mercy of an opaque LICM decision: an f32 reduction-heavy step (`gpt2_mini`, or the `scalar_rmw_accumulation` shape scaled up), A/B on the gate, per-kernel medians. A win there is attributable; a null result on cc would only say LLVM had already done it.

Note gh-569's warning applies to any follow-on tuning: this device's block-count curve is non-monotone — measure, do not maximize.

**Comment by lukstafi** [comment]
Decision (issue-wave gate, 2026-08-22; recommendation, veto if you disagree): (1) **relax the identity-precision gate unconditionally**, keeping the existing declines — the rewrite is already value-neutral at storage precision, and a config key would just be another unmeasured switch; (2) **the volatile shadow must not fire on the scope form** — the RMW it protects is gone at a localized site, and a volatile store-once is pure cost. Acceptance is the Metal per-kernel-median A/B on gpt2_mini; a cc null result only says LICM had already done it. Sequenced after #681 (the virtualizer rejects materialized scopes, so this localization lives only in codegen's peel).

**Comment by lukstafi** [comment]
PR up: lukstafi/ocannl-staging#421.

Both halves of the decision are in. (1) The identity-precision gate in `try_widen_serial_reduce` is gone unconditionally — renamed `try_localize_serial_reduce`, since residency is now the primary property and gh-639's widening is the special case. Every existing decline stays (debug logging, RNG-mentioning updates, `peel_accum_nest`'s structural refusals). (2) The volatile shadow does not fire on the scope form, and **no code change was needed for that** — worth recording, because the obvious fix would have been wrong. The shadow keys on the emitted `Set` reading its own node at a cell invariant across an enclosing serial loop; `peel_accum_nest` runs outermost-first, so localization lifts the `Set` out of exactly those loops and the predicate is already false. Where the peel is blocked at an outer level the store stays inside an invariant-address loop and the shadow still fires — correctly, since the per-iteration RMW is genuinely still there. Teaching `reads_tn` to ignore a `Local_scope` body's reads would have dropped protection at those partially-localized sites silently.

**The Metal measurement is much larger than this issue anticipated.** gpt2_mini forward, M4 Max, default schedule, arms alternated run by run, three reps, cache wiped between arms:

| | base | new | delta |
|---|---|---|---|
| sum of per-segment medians (117 segments) | 371.0 ms | 100.6 ms | −72.9% |
| whole step, median of 9 | 365.1 ms | 93.7 ms | −74.3% |
| `bench_gpt` step p50, 20 timed steps | 367.1 ms | 93.9 ms | −74.4% |
| `device volatile` RMW shadows in the emitted fused kernel | 124 | 0 | |

Per kernel the win is in the **contractions**, not the loss reduction: lm_head 32.99 → 6.51 ms (−80.3%), each of the eight FFN matmuls ~24.6 → ~5.5 ms (−77 to −78%), each attention projection ~6.2 → ~1.5 ms. `cross_entropy_loss` itself was only 0.211 ms and moved to 0.199 ms (−5.4%), inside the noise floor (~±0.02 ms, measured on zero-out kernels whose emitted source is byte-identical between arms).

Parity: the eight batch losses are bitwise identical in 7 of 8, the eighth differing by 5e-7 relative — the qualifier had also been blocking FMA contraction, so one ULP of reassociation is the only value change in the model.

Two things the issue asked about that the measurement could not reach: there is no per-kernel *median* instrument on Metal in-tree (`gpt2_kernel_harness.py` is HIP-only), so the per-segment numbers are min-of-20 within a run, median across three runs; and there is no batch-gradient kernel in gpt2_mini, which is `mode: infer` and whose diag runner builds `Train.forward` only — that shape is pinned by the row-reduction leg of the new test instead.
